whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,398 @@
+  some :: f a -> f [a]
+  many :: f a -> f [a]
+  {-# MINIMAL empty, (<|>) #-}
+  	-- Defined in ‘GHC.Base’
+instance Alternative ZipList -- Defined in ‘Control.Applicative’
+instance Alternative [] -- Defined in ‘GHC.Base’
+instance Alternative Maybe -- Defined in ‘GHC.Base’
+instance Alternative IO -- Defined in ‘GHC.Base’
+```
+
+This already gives us alot of useful information about that typeclass:
+
+1. A type implementing `Alternative` must be of kind `* -> *`
+2. It must also implement `Applicative` (and with that also `Functor`)
+3. To implement it you must implement `<|>` and `empty`
+4. Next to those 2 functions it gives you `some` and `many` as well
+5. Next to others it is also defined for the `Maybe` type we utilize in the example above.
+
+So lets take a look on the implementation of `Alternative Maybe`:
+
+```haskell
+instance Alternative Maybe where
+    empty = Nothing
+    Nothing <|> r = r
+    l       <|> _ = l
+```
+
+If you take a look at the definitions you can see that `empty :: f a` is
+the identity of `<|>` (very much like `mempty :: a` is the identity of
+`mappend :: a -> a -> a` in case of a `Monoid`. `Alternative` is also
+said to be _A monoid on applicative functors_.
+
+The truth table look like this.
+
+|             | **Nothing** | **Just r** |
+|-------------|-------------|------------|
+| **Nothing** | Nothing     | Just r     |
+| **Just l**  | Just l      | Just l     |
+
+
+Another propably interesting instance of instance of `Alternative` is `IO`.
+Here the first alternative from left to right which succeeds wins, f.e.
+
+```haskell
+test :: IO String
+test = fail "foo" <|> pure "bar" <|> fail "baz"
+```
+
+When executed the fist computation `fail "foo"` obvioudly fails, so the second one `pure "bar"` succeeds, which is why the third `fail "baz"` isn't executed at all:
+
+```none
+λ> test
+"bar"
+```
+
+## A simple Parser
+
+A very common usecase of the `Alternative` typeclass is in parsers.
+
+Lets asume we want to implement a simple Parser which reads a stream of 'i' and possibly returns a value of type `a` if the input matches.
+
+```haskell
+newtype Parser i a
+  Parser { runParser :: [i] -> (Maybe a, [i]) }
+
+evalParser :: (Show i) => Parser i a -> [i] -> Either String a
+evalParser p is = case runParser p is of
+                    (Nothing, is) -> Left $ "parser failed on " ++ show is
+                    (Just a, _)   -> Right a
+```
+
+This `Parser` type wraps a function which takes an input an gives back a `Just a` in case it matches or Nothing if not, together with ther remaining unused input.
+
+`evalParser` runs such a parser and returns a more convenient `Either` with
+an error string or the result.
+
+
+An implementation of a parser could be one which consumes a numeric input
+and succeeds if it smaller then a given number of fails if it is equal
+or greater that number:
+
+
+```haskell
+parseLT :: Ord a => a -> Parser a a
+parseLT a = Parser f
+    where f input@(i : is)
```

Comment:
This is a type application, right? At what point in their Haskell journey do we expect people to just know it? 😅

Reply:
Given that InstanceSigs is explained, it seems odd to not explain this one, if I understand correctly.

Reply:
You mean the `@`? 

No, thats not a type application but the name for the whole list (while `i` refers to the first element and `is` to the remaining so `input = i : is`.

Use that here because I want to:
1. Pattern match on the non empty list 
2. Have the `i` on hand for the 1st guard (` i < a     = (Just i, is)`)
3. But In the other guard case I need the whole list (`| otherwise = (Nothing, input)`). 
    Without the `input@` I could have written it as `| otherwise = (Nothing, i:is)`
    
I guess thats pretty standard haskell syntax which does not require any language extension.

Reply:
Ah, "as pattern". Cool, learned something new today, thanks!